### PR TITLE
hasOne association could not restore

### PR DESCRIPTION
**Pull request by JeskTop**
When I find @pobocks fix the problem https://github.com/radar/paranoia/pull/119 , and I use the new paranoia and find another bug. 
The problem can look at my first commit what is a test and the problem like https://github.com/radar/paranoia/issues/120 .

### CODE REVIEW

File: lib/paranoia.rb [hunk 1]
```diff
@@ -114,6 +114,10 @@ def restore_associated_records
           end
         end
       end
+
+      if association_data.nil? && association.macro.to_s == "has_one"
+        Object.const_get(association.name.to_s.camelize).only_deleted.where("#{self.class.name.to_s.underscore}_id", self.id).first.try(:restore, recursive: true)
```
Comment (radar) [inline lib/paranoia.rb]: Um, wow.

Could you please write some documentation about this? There's a lot of meta programming that goes on within this one line!

**Comment by JeskTop** [comment]
My commits fix the problem is the `has_one` counld not restore. 
I submitted before the test is passed, because the destroy object did not find from the database and it just a catch or not. Code: https://github.com/radar/paranoia/blob/rails4/test/paranoia_test.rb#L382

**Comment by JeskTop** [comment]
How can I find the has_one object?

The difference between `has_one` and `has_many` like this:

``` ruby
# User has_many Message
@user.messages.where(title: "abc")   #This is can work!
```

``` ruby
# User has_one Role
@user.role.where(name: "abc")  #This is can not work!
```

So if I want to find the `role` must use:

``` ruby
Role.where(name: "abc", user_id: @user.id)
```

It just like:

``` ruby
Object.const_get("role").where(name: "abc", user_id: @user.id)
```

So this one line:

``` ruby
Object.const_get(association.name.to_s.camelize).only_deleted.where("#{self.class.name.to_s.underscore}_id", self.id).first.try(:restore, recursive: true)
```

This mean is use `association.name.to_s.camelize` find the `has_one` model name and use `self.class.name.to_s.underscore` to setting this class `id` (`user_id`).

And then user `User has_one Role` it will work look like:

``` ruby
Object.const_get("role").only_deleted.where("user_id", self.id).first.try(:restore, recursive: true)
```

Or

``` ruby
Role.only_deleted.where("user_id", self.id).first.try(:restore, recursive: true)
```

File: lib/paranoia.rb (hunk 1 above)
Comment (JeskTop) [inline lib/paranoia.rb]: Hi @radar , my English is not very good, write to the following several documentations, do not know whether or not clear?

**Comment by pobocks** [comment]
Is this guaranteed to be the fkey name in question?

``` ruby
{self.class.name.to_s.underscore}_id
```

It seems like there'd need to be a check for key columns specified in the options.

**Comment by JeskTop** [comment]
@pobocks Oh. Yes! I must think about how to use the `foreign_key` to find the model.

**Comment by JeskTop** [comment]
Hey, @pobocks. I pushed the new commit and checked the specified `foreign_key` and `class_name`.

**Comment by adityashedge** [comment]
I'm having an issue with restoring dependent has_one relation recursively in rails 4.1. Can you check if this merge request fixes it and if it does please merge it ASAP. Thank  you in advance.

Exact Problem:
User has_one UserInfo and UserInfo belongs_to User.
Unable to restore UserInfo when restoring user and all dependent records recursively.
UserInfo is soft deleted but not restored when restoring User recursively.

**Comment by JeskTop** [comment]
@adityashedge 
Thanks. I think the commit can fix all of your problems.
You can see my test. If `User has_one UserInfo and UserInfo belongs_to User`, It will restore `UserInfo` when restoring user and all dependent records recursively.
And if you find the another problem, you can tell me.

**Comment by willsu** [comment]
+1 @JeskTop This fixed my issue with `restore(recursive: true)` on a has_one association.

@radar Any chance of getting this pull request accepted? I'm pointing my Gemfile at the @JeskTop fork and `setting_restore` branch for the time being.

**Comment by radar** [comment]
The PR no longer applies cleanly to the targeted branch. If this PR got an update to fix that, then I would gladly pull this in.

**Comment by JeskTop** [comment]
What do you mean clean? I alreay update the `rails4` branch and merge for my `setting_restore` branch. If @radar has another problems, you can tell me and I will fix that. Thanks.

**Comment by ccaratori** [comment]
Hey guys is this going to be merged and released any time soon?

**Comment by radar** [comment]
@JeskTop I still can't automatically merge this PR, and it has failing tests.

screen shot 2014-09-15 at 11 18 38 am

Once those failing tests and merge conflicts are fixed, then I can look at this again.

**Comment by JeskTop** [comment]
@radar Thanks. I already fix all conflicts. And push the commit.

**Comment by radar** [comment]
Thank you very much @JeskTop.

**Pull request merged**